### PR TITLE
"Reaching Definitions" in support of identifying used-but-not-set scripting issues

### PR DESCRIPTION
**Pull request by vpax**
This next set of script optimization changes introduces classes to compute "reaching definitions" (RDs), i.e., for script variables, and, potentially, record fields, tracking the extent to which a given definition (assignment) can be visible elsewhere in the script.  The analysis works on local variables in function/event/hook bodies (but not across them).

The code tracks two forms of RDs, "minimal" (what's guaranteed to reach a given point in a function body) and "maximal" (what possibly could reach).  For now, we only actually leverage the former, but some upcoming script optimization will use the latter.

Minimal RDs allow identification of places for which a value is used where it does not appear that it will necessarily be defined.  Specifying the `-u` option will generate warnings for instances where this holds for local variables.  Specifying `-uu` turns on additional (expensive) analysis to report instances where record fields might be used without having previously been set.

The base scripts have some places where the static analysis lacks sufficient power to tell that values are being used safely (guaranteed to have been identified).  In order to enable users to employ `-u` on their own scripts without being distracted by these instances, this change also includes a new attribute, `&is_set`, which can be associated with a variable or a record field to inform Zeek's analysis that the script writer asserts the value will be set, suppressing the associated warnings.

I've structured the commits as a number of self-contained changes (each of them builds without error).  The most complex is https://github.com/zeek/zeek/commit/2f0d02c878b90f19de9abd86c25337be1db6437d near the end; it introduces the class that walks an AST and does the work of propagating RDs and identifying used-but-not-set hazards.

For testing, there's now a new `-a usage` alternative that addresses some minor differences, and a new test, `language/uninitialized-local3.zeek`, that exercises both `-uu` and the `&is_set` functionality. The very last commit is actually already needed by the recently introduced `-a xform` test suite alternative (due to an even more recent test suite change).  It's included here because `-a usage` builds on that alternative, so also runs into the same issue.

### CODE REVIEW

**Review by jsiwek — APPROVED** [review]
@vpax Thanks, everything looks good to go if you want to check whether any of my changes would be disruptive.  Also:

* The use of "set" in `&is_set` has some overload with the meaning of `set` the data type.  I'd be fine leaving it, but wonder if there's enough interest from anyone to use something more intuitive.  Maybe `&is_assigned` or `&no_usage_warning`.

* Checking my understanding about `&is_set` when used on a record-declaration field: is that expected to get used only rarely or in what cases would someone want to do that?  I couldn't think of when I'd ever want to use it since that pre-decides for any arbitrary user/usage that "it's fine" when really that doesn't seem like my call to make.

**Comment by vpax** [comment]
@jsiwek thanks a bunch, changes look good.  Re `&is_set`, I see your point.  The next set of changes are going to introduced an analogous is-set-but-not-used (which I currently turn off via `&is_used`).  So we can either have two such attributes, that one and `&is_assigned`, or double them up into one (`&no_usage_warning`) at the loss of knowing just what about the usage might appear wonky.  I guess I'm inclined to the former given it's a bit more clear, but I don't feel at all strongly.

Regarding `&is_set` and record fields, one possible use I had in mind is where the event engine guarantees the field will be set even though the control flow analysis can't infer that.  That said, I don't believe that case came up as an issue for any of the base scripts or the test suite, so we could drop it for now and only add it back when we encounter such issues.

File: src/script_opt/GenRDs.cc
Comment (vpax) [reply src/script_opt/GenRDs.cc:974]: Thanks for flagging, let me look into this.  I know there are set-but-not-used issues that inlining can create (because of the opportunity to do constant propagation across the inlining, and so prune some branches that have the actual uses), but am not remembering offhand if there's something similar for used-but-not-set.  Will follow up once I've tried this out.

Comment (vpax) [reply src/script_opt/GenRDs.cc]: This turned out to be fruitful to revisit.  In the process I found a number of bugs.  I've now resolved those, and also renamed the attribute to `&is_assigned` per discussion.  Back over to you, Jon.

**Comment by jsiwek** [comment]
@vpax made a couple more "quality" tweaks if you want to check them over, but otherwise ready to merge (if CI doesn't tell me I did something dumb)

* For the `RDMap()` accessor simplification I did: simpler to always return const-reference and let caller decide if they want to copy (with copy meaning +1 `shared_ptr` ref-count).

* I noticed `ReachingDefs::AddRD()` had potential for a redundant map lookup and eliminating some of those shaved ~15%.  Think the `CopyMap()` still takes notable amount of time overall and might be a place to look for potential optimizations (for later/whenever).

**Review by vpax — COMMENTED** [review]
Looks good, thanks.

**Comment by vpax** [comment]
@jsiwek Your latest changes all look good.  I've pushed a tweak to the `language/uninitialized-local3.zeek` test case, as I noticed it didn't include the basic case of `&is_assigned` used for an entire variable rather than just a record field.  I also added a comment that after the printing of the uninitialized local, execution stops (due to `ExprError`), to help maintainers in the future not waste time (like I did!) adding code after that point.

**Pull request merged**
